plugins/plugins/casbin/filter.go

```diff
@@ -35,33 +35,27 @@ type filter struct {
 	config    *config
 }
 
+var (
+	Changed   = false
+	ChangedMu sync.RWMutex
+)
+
 func (f *filter) DecodeHeaders(headers api.RequestHeaderMap, endStream bool) api.ResultAction {
 	conf := f.config
 	role, _ := headers.Get(conf.Token.Name) // role can be ""
 	url := headers.Url()
 
-	policyChanged := file.IsChanged(conf.modelFile, conf.policyFile)
-	if policyChanged && !conf.updating.Load() {
-		conf.updating.Store(true)
-		api.LogWarnf("policy %s or model %s changed, reload enforcer", conf.policyFile.Name, conf.modelFile.Name)
+	err := file.WatchFiles(func() {
+		setChanged(true)
+	}, conf.modelFile, conf.policyFile)
 
-		go func() {
-			defer conf.updating.Store(false)
-			defer f.callbacks.RecoverPanic()
-
-			e, err := casbin.NewEnforcer(conf.Rule.Model, conf.Rule.Policy)
-			if err != nil {
-				api.LogErrorf("failed to update Enforcer: %v", err)
-				// next request will retry
-			} else {
-				conf.lock.Lock()
-				conf.enforcer = e
-				conf.lock.Unlock()
+	if err != nil {
+		api.LogErrorf("failed to watch files: %v", err)
+		return &api.LocalResponse{Code: 500}
+	}
 
-				file.Update(conf.modelFile, conf.policyFile)
-				api.LogWarnf("policy %s or model %s changed, enforcer reloaded", conf.policyFile.Name, conf.modelFile.Name)
-			}
-		}()
+	if getChanged() {
+		conf.reloadEnforcer()
```

Comment:
Why don't we call this inside `file.WatchFiles(func() {`?

Reply:
because if ```reloadenforcer``` called in ```file.WatchFiles()```, the config won't be changed. So I need the flag

Reply:
I wonder why. Could you explain it?

Reply:
I don't know. In fact, this problem bother me for a long time.